Modules/timemodule.c

```diff
@@ -2066,23 +2070,44 @@ pysleep(_PyTime_t secs)
         return -1;
     }
     deadline = monotonic + secs;
+#if defined(HAVE_CLOCK_NANOSLEEP) && !defined(MS_WINDOWS)
+    if (_PyTime_AsTimespec(deadline, &timeout) < 0) {
+        return -1;
+    }
+#endif
 
     do {
 #ifndef MS_WINDOWS
-        if (_PyTime_AsTimeval(secs, &timeout, _PyTime_ROUND_CEILING) < 0)
+#ifndef HAVE_CLOCK_NANOSLEEP
+        if (_PyTime_AsTimeval(secs, &timeout, _PyTime_ROUND_CEILING) < 0) {
             return -1;
+        }
+#endif
 
         Py_BEGIN_ALLOW_THREADS
+#ifdef HAVE_CLOCK_NANOSLEEP
+        err = clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &timeout, NULL);
+#else
         err = select(0, (fd_set *)0, (fd_set *)0, (fd_set *)0, &timeout);
+#endif
         Py_END_ALLOW_THREADS
 
-        if (err == 0)
+        if (err == 0) {
             break;
+        }
 
+#ifdef HAVE_CLOCK_NANOSLEEP
+        if (err != EINTR) {
+            errno = err;
```

Comment:
Can you merge the two if/else code  paths? Only this line is different if HAVE_CLOCK_NANOSLEEP is defined.

Reply:
In case of using select(), errno must be compared in if statement. If clock_nanosleep() is used its (err) return value must be in the if statement comparing. If i am not following this, some unit tests will be failed because there are any magic behavier/differences in errorno and the returned err of clock_nanosleep() and select().

Reply:
I put errno value passing between Py_BEGIN_ALLOW_THREADS and Py_END_ALLOW_THREADS in HAVE_CLOCK_NANOSLEEP. If it is allowed in this block, it can merging nicely.